os/arch/arm/src/armv7-m/up_assert.c

```diff
@@ -250,95 +251,103 @@ static void up_dumpstate(void)
 
-	/* Does the current stack pointer lie within the interrupt
-	 * stack?
+	/* Check if the assert location is in user thread or IRQ handler.
+	 * If the irq_num is lesser than NVIC_IRQ_USAGEFAULT, then it is
+	 * a fault and not an irq.
 	 */
-
-	if (sp <= istackbase && sp > istackbase - istacksize) {
-		/* Yes.. dump the interrupt stack */
-
-		up_stackdump(sp, istackbase);
+	if (g_irq_nums[2] || (g_irq_nums[1] && g_irq_nums[0] > NVIC_IRQ_USAGEFAULT)) {
+		/* Assert in nested irq */
+		isIrqAssert = true;
+		lldbg("Code asserted in nested IRQ state!\n");
+	} else if ((g_irq_nums[1] && (g_irq_nums[0] <= NVIC_IRQ_USAGEFAULT)) || (g_irq_nums[0] > NVIC_IRQ_USAGEFAULT)) {
+		/* Assert in irq */
+		isIrqAssert = true;
+		lldbg("Code asserted in IRQ state!\n");
+	} else {
+		/* Assert in user thread */
+		lldbg("Code asserted in normal thread!\n");
+		if (current_regs) {
+			/* If assert is in user thread, but current_regs is not NULL,
+			 * it means that assert happened due to a fault. So, we want to
+			 * reset the sp to the value just before the fault happened
+			 */
+			sp = current_regs[REG_R13];
+		}
 	}
 
-	/* Extract the user stack pointer if we are in an interrupt handler.
-	 * If we are not in an interrupt handler.  Then sp is the user stack
-	 * pointer (and the above range check should have failed).
-	 */
-
-	if (current_regs) {
-		sp = current_regs[REG_R13];
-		lldbg("sp:     %08x\n", sp);
-	}
+	/* Print IRQ handler details if required */
 
-	lldbg("User stack:\n");
-	lldbg("  base: %08x\n", ustackbase);
-	lldbg("  size: %08x\n", ustacksize);
-#ifdef CONFIG_STACK_COLORATION
-	lldbg("  used: %08x\n", up_check_tcbstack(rtcb));
+	if (isIrqAssert) {
+		uint8_t crashIrqNum = 1;
+		if (g_irq_nums[1] == 0 && g_irq_nums[0] > NVIC_IRQ_USAGEFAULT) {
+			crashIrqNum = 0;
+		}
+		lldbg("IRQ num: %d\n", g_irq_nums[crashIrqNum]);
+		lldbg("IRQ handler: %08x \n", g_irqvector[g_irq_nums[crashIrqNum]].handler);
+#ifdef CONFIG_DEBUG_IRQ_INFO
+		lldbg("IRQ name: %s \n", g_irqvector[g_irq_nums[crashIrqNum]].irq_name);
 #endif
-
-	/* Dump the user stack if the stack pointer lies within the allocated user
-	 * stack memory.
-	 */
-
-	if (sp <= ustackbase && sp > ustackbase - ustacksize) {
-		up_stackdump(sp, ustackbase);
-	}
-#else
-
-	/* Show user stack info */
-
-	lldbg("sp:         %08x\n", sp);
-	lldbg("stack base: %08x\n", ustackbase);
-	lldbg("stack size: %08x\n", ustacksize);
-#ifdef CONFIG_STACK_COLORATION
-	lldbg("stack used: %08x\n", up_check_tcbstack(rtcb));
+#if CONFIG_ARCH_INTERRUPTSTACK > 3
+		if ((sp <= istackbase) && (sp > (istackbase - istacksize))) {
+			stackbase = istackbase;
+			stacksize = istacksize;
+			lldbg("Current SP is IRQ SP: %08x\n", sp);
+			lldbg("IRQ stack:\n");
+		} else
 #endif
+		{
+			isSpCorrupt = true;
+		}
+	} else if ((sp <= stackbase) && (sp > (stackbase - stacksize))) {
+		lldbg("Current SP is User Thread SP: %08x\n", sp);
+		lldbg("User stack:\n");
+	} else {
+		isSpCorrupt = true;
+	}
```

Comment:
This looks strange. Could you check the code without CONFIG_ARCH_INTERRUPTSTACK?